Evolution/0016-mutli-producer-single-consumer-channel.md

```diff
@@ -0,0 +1,652 @@
+are the behaviors where `Async[Throwing]Stream` diverges from the expectations.
+
+- Backpressure: Doesn't expose a "resumption" signal to the producer
+- Multi/single consumer:
+  - Divergent implementation between throwing and non-throwing variant
+  - Supports multiple consumers even though proposal positions it as a unicast
+  asynchronous sequence
+- Consumer termination: Doesn't handle the `Continuation` being `deinit`ed
+- Producer termination: Happens on first consumer termination 
+
+This section proposes a new type called `MutliProducerSingleConsumerChannel` that implement all of
+the above-mentioned behaviors.
+
+### Creating an MutliProducerSingleConsumerChannel
+
+You can create an `MutliProducerSingleConsumerChannel` instance using the new
+`makeChannel(of: backpressureStrategy:)` method. This method returns you the
+channel and the source. The source can be used to send new values to the
+asynchronous channel. The new API specifically provides a
+multi-producer/single-consumer pattern.
+
+```swift
+let (channel, source) = MutliProducerSingleConsumerChannel.makeChannel(
+    of: Int.self,
+    backpressureStrategy: .watermark(low: 2, high: 4)
+)
+```
+
+The new proposed APIs offer three different ways to bridge a backpressured
+system. The foundation is the multi-step synchronous interface. Below is an
+example of how it can be used:
+
+```swift
+do {
+    let sendResult = try source.send(contentsOf: sequence)
+    
+    switch sendResult {
+    case .produceMore:
+       // Trigger more production
+    
+    case .enqueueCallback(let callbackToken):
+        source.enqueueCallback(token: callbackToken, onProduceMore: { result in
+            switch result {
+            case .success:
+                // Trigger more production
+            case .failure(let error):
+                // Terminate the underlying producer
+            }
+        })
+    }
+} catch {
+    // `send(contentsOf:)` throws if the asynchronous stream already terminated
+}
+```
+
+The above API offers the most control and highest performance when bridging a
+synchronous producer to an asynchronous sequence. First, you have to send
+values using the `send(contentsOf:)` which returns a `SendResult`. The result
+either indicates that more values should be produced or that a callback should
+be enqueued by calling the `enqueueCallback(callbackToken: onProduceMore:)`
+method. This callback is invoked once the backpressure strategy decided that
+more values should be produced. This API aims to offer the most flexibility with
+the greatest performance. The callback only has to be allocated in the case
+where the producer needs to be suspended.
+
+Additionally, the above API is the building block for some higher-level and
+easier-to-use APIs to send values to the channel. Below is an
+example of the two higher-level APIs.
+
+```swift
+// Writing new values and providing a callback when to produce more
+try source.send(contentsOf: sequence, onProduceMore: { result in
```

Comment:
can this be written without ambiguity as a trailing closure?

Reply:
Yes in a synchronous context both `send` methods can be called without ambiguity, e.g. this compiles:
```
try source.send(1)
source.send(2) { result in
    print(result)
}
```

The only thing that needs some special handling is if you want to call a sync send from an async context since the async methods are preferred; however, users can workaround this by creating an inline closure, e.g.:

```
try { try source.send(1) }()
```